plutus-core/plutus-core/src/PlutusCore/Builtin/KnownTypeAst.hs

```diff
@@ -143,77 +150,89 @@ data family RepHole x
 type TypeHole :: forall hole. GHC.Type -> hole
 data family TypeHole a
 
--- | For annotating an uninstantiated built-in type, so that it gets handled by the right instance
--- or type family.
-type BuiltinHead :: forall k. k -> k
-data family BuiltinHead f
-
--- | Take an iterated application of a built-in type and elaborate every function application
--- inside of it to 'TypeAppRep', plus annotate the head with 'BuiltinHead'.
--- The idea is that we don't need to process built-in types manually if we simply add some
--- annotations for instance resolution to look for. Think what we'd have to do manually for, say,
--- 'ToHoles': traverse the spine of the application and collect all the holes into a list, which is
--- troubling, because type applications are left-nested and lists are right-nested, so we'd have to
--- use accumulators or an explicit 'Reverse' type family. And then we also have 'KnownTypeAst' and
--- 'ToBinds', so handling built-in types in a special way for each of those would be a hassle,
--- especially given the fact that type-level Haskell is not exactly good at computing things.
--- With the 'ElaborateBuiltin' approach we get 'KnownTypeAst', 'ToHoles' and 'ToBinds' for free.
-type ElaborateBuiltin :: forall k. k -> k
-type family ElaborateBuiltin a where
-    ElaborateBuiltin (f x) = ElaborateBuiltin f `TyAppRep` x
-    ElaborateBuiltin f     = BuiltinHead f
+-- | Specifies that the given type is a built-in one and can be embedded into a 'Type'.
+type HasTypeLevel :: forall a. (GHC.Type -> GHC.Type) -> a -> GHC.Constraint
+type HasTypeLevel uni x = KnownTypeAst Void uni (ElaborateBuiltin uni x)
```

Comment:
I guess this could instead be quantified over the type name argument?

Reply:
Yes, but this what I'm actively avoiding here, because see the [other](https://github.com/input-output-hk/plutus/pull/5434#discussion_r1272288449) comment.

Reply:
Is it true that ultimately the only reason that you ever need `HasTypeLevel` is to allow you to eventually call `mkTyBuiltin` and you only need `HasTermLevel` if you're eventually going to call `mkConstant`?  If that's the case then I hadn't grasped it before (maybe spittling the constraints has clarified things), so it might be worth mentioning it here.

Reply:
> Is it true that the only reason that you ever need `HasTypeLevel` is to allow you to eventually call `mkTyBuiltin` and you only need `HasTermLevel` if you're eventually going to call `mkConstant`?

I think this is correct, yes. Not necessarily `mkConstant`, can be `someValue` for example, but that's what `mkConstant` is defined in terms of.

>  If that's the case then I hadn't grasped it before (maybe spittling the constraints has clarified things), so it might be worth mentioning it here.

I'm not sure about that... What if somebody decides they want to call `toTypeAst` directly given the `HasTypeLevel` constraint? They might even decide to call `toTypeAst` over the argument of the built-in type in the constraint rather than the whole built-in type for example. 

Let's return to this question once we split the universe and everything regarding the design of builtins is finalized.